### PR TITLE
[Feature] [Flink-K8s-V2] Refactor the state tracking of Flink on Kubernetes

### PR DESCRIPTION
**Pull request by ChengJie1053**
Refactor the state tracking of Flink on Kubernetes

issues:https://github.com/apache/incubator-streampark/issues/2885

### CODE REVIEW

File: streampark-console/streampark-console-service/src/main/scala/org/apache/streampark/console/core/task/FlinkK8sChangeListener.scala [hunk 1]
```diff
@@ -0,0 +1,196 @@
+package org.apache.streampark.console.core.task
```
Comment (wolfboys) [inline streampark-console/streampark-console-service/src/main/scala/org/apache/streampark/console/core/task/FlinkK8sChangeListener.scala:18]: Missing  license header

Comment (ChengJie1053) [reply streampark-console/streampark-console-service/src/main/scala/org/apache/streampark/console/core/task/FlinkK8sChangeListener.scala:18]: Ok, thanks for reviewing the code

**Comment by wolfboys** [comment]
cc@Al-assad  PTAL

File: streampark-console/streampark-console-service/src/main/scala/org/apache/streampark/console/core/task/FlinkK8sChangeListener.scala [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        else 0)
+    }
+
+    app.setState(state.getValue)
+    // when a flink job status change event can be received, it means
+    // that the operation command sent by streampark has been completed.
+    app.setOptionState(OptionState.NONE.getValue)
+  }
+
+  def inferEvalJobStateFromPersist(current: EvalJobState, previous: EvalJobState): EvalJobState = {
```
Comment (Al-assad) [inline streampark-console/streampark-console-service/src/main/scala/org/apache/streampark/console/core/task/FlinkK8sChangeListener.scala]: No need to infer the current job status from persistent storage, directly use the job status obtained from the FlinkK8sObserver.evaluatedJobSnaps subscription.

File: streampark-console/streampark-console-service/src/main/scala/org/apache/streampark/console/core/task/FlinkK8sChangeListener.scala [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+      .foreach {
+        jobSnap =>
+          ZIO
+            .attempt {
+              // update application recrod
+              val appId = jobSnap.appId
+              // get pre application record
+              val app: Application = applicationService.getById(appId)
+              if (app == null || jobSnap.evalState != null) return ZIO.unit
+              applicationService.persistMetrics(app)
```
Comment (Al-assad) [inline streampark-console/streampark-console-service/src/main/scala/org/apache/streampark/console/core/task/FlinkK8sChangeListener.scala]: exchange line 77 and 78

File: streampark-console/streampark-console-service/src/main/scala/org/apache/streampark/console/core/task/FlinkK8sChangeListener.scala [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                    trackedKey.clusterNamespace == namespaceAndName._1 && trackedKey.clusterName == namespaceAndName._2)
+                .someOrFail(TrackKeyNotFound(namespaceAndName._1, namespaceAndName._2))
+              val restSvcEndpoint: RestSvcEndpoint = restSvcEndpointSnap._2
+
+              val app: Application = applicationService.getById(trackKey.map(_.id))
+
+              val flinkCluster: FlinkCluster = flinkClusterService.getById(app.getFlinkClusterId)
+
+              if (restSvcEndpoint == null || restSvcEndpoint.chooseRest == null) return ZIO.unit
+              val url = restSvcEndpoint.chooseRest
```
Comment (Al-assad) [inline streampark-console/streampark-console-service/src/main/scala/org/apache/streampark/console/core/task/FlinkK8sChangeListener.scala]: Trends to use restSvcEndpoint.ipRest here

**Review by Al-assad — CHANGES_REQUESTED** [review]
Need to delete org/apache/streampark/console/core/task/FlinkK8sChangeEventListener.java because it will generate a class with the same fully qualified name after compilation.

**Comment by Al-assad** [comment]
@ChengJie1053 A small tip, it can use this style to consume ZStream safely 😃：

```scala
  private val alterStateList =
    Array(FlinkAppState.FAILED, FlinkAppState.LOST, FlinkAppState.RESTARTING, FlinkAppState.FINISHED)

  def subscribeJobStatusChange: UIO[Unit] = {
    FlinkK8sObserver.evaluatedJobSnaps
      .flatSubscribeValues()
      // Get Application records and convert JobSnapshot to Application
      .mapZIO { jobSnap =>
        ZIO
          .attemptBlocking {
            Option(applicationService.getById(jobSnap.appId))
              .map(app => setByJobStatusCV(app, jobSnap))
          }
          .catchAll { err =>
            logError(s"Fail to get Application records: ${err.getMessage}")
              .as(None) @@ annotated("appId" -> jobSnap.appId.toString)
          }
      }
      .filter(_.nonEmpty)
      .map(_.get)
      // Save Application records
      .tap { app =>
        ZIO
          .attemptBlocking(applicationService.persistMetrics(app))
          .retryN(3)
          .tapError(err => logError(s"Fail to persist Application status: ${err.getMessage}"))
          .ignore @@ annotated("appId" -> app.getAppId)
      }
      // Alert for unhealthy states in parallel
      .mapZIOPar(10) { app =>
        val state = FlinkAppState.of(app.getState)
        ZIO
          .attemptBlocking(alertService.alert(app, state))
          .when(alterStateList.contains(state))
          .retryN(3)
          .tapError(err => logError(s"Fail to alter unhealthy application state: ${err.getMessage}"))
          .ignore @@ annotated("appId" -> app.getAppId, "state" -> state.toString)
      }
      .runDrain
  }
```

File: streampark-console/streampark-console-service/src/main/scala/org/apache/streampark/console/core/task/FlinkK8sChangeListener.scala (hunk 4 above)
Comment (ChengJie1053) [reply streampark-console/streampark-console-service/src/main/scala/org/apache/streampark/console/core/task/FlinkK8sChangeListener.scala]: Ok, thanks for reviewing the code

**Comment by ChengJie1053** [comment]
> 

Ok, thank you

**Comment by ChengJie1053** [comment]
> @ChengJie1053 A small tip, it can use this style to consume ZStream safely 😃：
> 
> ```scala
>   private val alterStateList =
>     Array(FlinkAppState.FAILED, FlinkAppState.LOST, FlinkAppState.RESTARTING, FlinkAppState.FINISHED)
> 
>   def subscribeJobStatusChange: UIO[Unit] = {
>     FlinkK8sObserver.evaluatedJobSnaps
>       .flatSubscribeValues()
>       // Get Application records and convert JobSnapshot to Application
>       .mapZIO { jobSnap =>
>         ZIO
>           .attemptBlocking {
>             Option(applicationService.getById(jobSnap.appId))
>               .map(app => setByJobStatusCV(app, jobSnap))
>           }
>           .catchAll { err =>
>             logError(s"Fail to get Application records: ${err.getMessage}")
>               .as(None) @@ annotated("appId" -> jobSnap.appId.toString)
>           }
>       }
>       .filter(_.nonEmpty)
>       .map(_.get)
>       // Save Application records
>       .tap { app =>
>         ZIO
>           .attemptBlocking(applicationService.persistMetrics(app))
>           .retryN(3)
>           .tapError(err => logError(s"Fail to persist Application status: ${err.getMessage}"))
>           .ignore @@ annotated("appId" -> app.getAppId)
>       }
>       // Alert for unhealthy states in parallel
>       .mapZIOPar(10) { app =>
>         val state = FlinkAppState.of(app.getState)
>         ZIO
>           .attemptBlocking(alertService.alert(app, state))
>           .when(alterStateList.contains(state))
>           .retryN(3)
>           .tapError(err => logError(s"Fail to alter unhealthy application state: ${err.getMessage}"))
>           .ignore @@ annotated("appId" -> app.getAppId, "state" -> state.toString)
>       }
>       .runDrain
>   }
> ```

Ok, thank you

**Review by Al-assad — DISMISSED** [review]
LGTM, Thank for your contribution, I will submit a pull request later to improve some minor code style issues.

**Comment by Al-assad** [comment]
@ChengJie1053 Oh, there are some code conflicts at 
image

**Review by Al-assad — APPROVED** [review]

**Comment by ChengJie1053** [comment]
> LGTM, Thank for your contribution, I will submit a pull request later to improve some minor code style issues.

Ok, thanks for reviewing the code

**Pull request merged**